### ISSUE TITLE
_sample_center_embeddings

### ISSUE THREAD

**Issue by powermano** [iamsangmeng/InterFace#2]
I have modified the way of getting _sample_center_embeddings and thought it was equivalent, but the loss gets NAN after a few iters.
It makes me confused. My changed code is as following:
```
        _sample_center_embeddings = torch.zeros((labels.shape[0], self.embedding_size), dtype=self.weight_activated.dtype, device=self.weight_activated.device)

        idx = torch.where(labels != -1)[0]

        _sample_center_embeddings[idx] = self.weight_activated[labels[idx].view(-1)]

        distributed.all_reduce(_sample_center_embeddings, op=distributed.ReduceOp.SUM)
```

And the whole forward  fuction in partial_fc is :
```
 """
        Parameters:
        ----------
        local_embeddings: torch.Tensor
            feature embeddings on each GPU(Rank).
        local_labels: torch.Tensor
            labels on each GPU(Rank).

        Returns:
        -------
        loss: torch.Tensor
            pass
        """
        local_labels.squeeze_()
        local_labels = local_labels.long()
        self.update()

        batch_size = local_embeddings.size(0)
        if self.last_batch_size == 0:
            self.last_batch_size = batch_size
        assert self.last_batch_size == batch_size, (
            "last batch size do not equal current batch size: {} vs {}".format(
                self.last_batch_size, batch_size))

        _gather_embeddings = [
            torch.zeros((batch_size, self.embedding_size)).cuda()
            for _ in range(self.world_size)
        ]
        _gather_labels = [
            torch.zeros(batch_size).long().cuda() for _ in range(self.world_size)
        ]
       
        _list_embeddings = AllGather(local_embeddings, *_gather_embeddings)
        distributed.all_gather(_gather_labels, local_labels)

        embeddings = torch.cat(_list_embeddings)
        labels = torch.cat(_gather_labels)
  

        labels = labels.view(-1, 1)
        index_positive = (self.class_start <= labels) & (
                labels < self.class_start + self.num_local
        )
        # print(self.rank, labels)

        labels[~index_positive] = -1

        labels[index_positive] -= self.class_start

        if self.sample_rate < 1:
            self.sample(labels, index_positive, optimizer)
        
        _sample_center_embeddings = torch.zeros((labels.shape[0], self.embedding_size), dtype=self.weight_activated.dtype, device=self.weight_activated.device)

        idx = torch.where(labels != -1)[0]

        _sample_center_embeddings[idx] = self.weight_activated[labels[idx].view(-1)]

        distributed.all_reduce(_sample_center_embeddings, op=distributed.ReduceOp.SUM)

        with torch.cuda.amp.autocast(self.fp16):
            norm_embeddings = normalize(embeddings) # (B*n, C)
            norm_sample_center_embeddings = normalize(_sample_center_embeddings) # (B*n, C)
            norm_weight_activated = normalize(self.weight_activated)
            logits = linear(norm_embeddings, norm_weight_activated)
            center_logits = linear(norm_sample_center_embeddings, norm_weight_activated)
        if self.fp16:
            logits = logits.float()
            center_logits = center_logits.float()

        logits = logits.clamp(-1, 1)
        center_logits = center_logits.clamp(-1, 1)

        logits = self.margin_softmax(logits, labels, center_logits)
        loss = self.dist_cross_entropy(logits, labels)
        return loss

```

**Comment by iamsangmeng** [comment]
I can understand what you want to rewrite, but you did not rewrite this part of the content, the main error is still in the `idx` variable changes, I hope you can patiently check.

**Comment by Meng-Sang** [comment]
You did not rewrite this part of the content, the main error is still in the `idx` variable changes, I hope you can patiently check.
